Redfish: enable device listing and the tentative update function

lcp opened the pull request:
This series of patches fixes some issues in the redfish plugin when enumerating the devices. I've tested the plugin in a HPE Gen10 machine, and at least "fwupdmgr get-devices" showed the devices from FirmwareInventory. However, the HPE Gen10 machine doesn't provide the service IP in SMBIOS, so I have to add a conf file for the service IP and username/password. I'm not sure if there is any other method to discover the IP.

I followed the HPE document(*) to implement update, but it's not guaranteed to work due to lack of a valid firmware update file for testing.

(*) https://github.com/HewlettPackard/ilo-rest-api-docs/blob/master/source/includes/_ilo5_updateservicedoc.md

superm1 commented:
As an obvious note, please clean up the CI failures.

Since at least in the implementation you tested the service IP isn't automatically discoverable, it may make sense to add some documentation either in a README.md, wiki page, or in the conf file itself on how one might discover this IP and port.  Pointers for common ports or IPs may be useful too.

superm1 commented on plugins/redfish/fu-redfish-client.c:
```diff
@@ -76,34 +101,62 @@ fu_redfish_client_coldplug_member (FuRedfishClient *self,
-	if (json_object_get_boolean_member (member, "Updateable"))
+	if (json_object_has_member (member, "Updateable")) {
+		if (json_object_get_boolean_member (member, "Updateable"))
+			fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_UPDATABLE);
```
Something we have done recently in another plugin was to set the update error state when the device isn't updatable for the reason it's not updatable.  For example if the reason that the json didn't have an Updateable member was because of a BIOS option for example that would be a good reason to advertise here in this situation.

lcp replied:
I've contacted the HPE engineer, and he said the Updateable field is not implemented yet, so I really don't know if they are updatable or not. If this could be a problem, I'll remove the bit if Updateable is absent.

superm1 commented on plugins/redfish/fu-plugin-redfish.c:
```diff
@@ -39,11 +51,13 @@ fu_plugin_startup (FuPlugin *plugin, GError **error)
+	redfish_uri = fu_plugin_get_config_value (plugin, "RedfishUri");
 	if (redfish_uri != NULL) {
+		g_autofree gchar *username;
+		g_autofree gchar *password;
```
These should be initialized to `NULL`.

lcp replied:
Will fix it.

lcp commented on plugins/redfish/fu-redfish-client.c:
```diff
@@ -76,34 +101,62 @@ fu_redfish_client_coldplug_member (FuRedfishClient *self,
+	if (json_object_has_member (member, "SoftwareId")) {
+		guid = json_object_get_string_member (member, "SoftwareId");
+	} else if (json_object_has_member (member, "Oem")) {
+		JsonObject *oem = json_object_get_object_member (member, "Oem");
```
Thanks for catching this. Will fix it.

lcp commented on plugins/redfish/fu-redfish-client.c:
```diff
@@ -438,11 +559,78 @@ fu_redfish_client_set_smbios_interfaces (FuRedfishClient *self,
+	/* create URI */
+	uri = soup_uri_new (NULL);
+	soup_uri_set_scheme (uri, g_strcmp0 (self->hostname, "localhost") == 0 ?
+				  "http" : "https");
```
I keep the code for the emulator. Maybe I should make the scheme be a part of RedfishUri in the config file.

superm1 replied:
Yeah that sounds good.

lcp commented:
I've fixed the errors and updated README.md. RedfishUri now requires the scheme (http or https) to be specified. I also added one more patch to relax the SSL check since the self-signed certificate in BMC usually cannot be used to verify the hostname.

hughsie commented:
Generally looks good, with a few nits. With the commit msg we generally do 'plugin: Sentancecase commit message' with again Sentancecase body. That should be easy to fix up. Also, sorry for the delay, just got back from holiday.
